File: crates/mumble-tauri/ui/src/core/systemBars.ts
/**
 * The phone's status and gesture bars, told what the page paints behind them.
 *
 * The app is drawn edge to edge on Android, so the page is what shows through
 * both bars - and the bars' icons stay white unless told otherwise, which over
 * a light theme is a clock and a battery nobody can see. The native side
 * (`SystemBarsPlugin`) flips them dark on request, and sends the page its
 * inset variables again with every answer, so the first call after a load is
 * also how the page learns where the bars are.
 *
 * Pack-neutral, like `windowIcon`: a pack says whether its surface is light.
 */
import { invoke } from "@tauri-apps/api/core";
import { isMobile } from "@core/utils/platform";

/** Dark bar icons over a light page, light ones over a dark page. */
export async function applySystemBarStyle(light: boolean): Promise<void> {
  // A desktop window has no system bars; the command answers Ok there, but
  // there is no reason to cross the IPC boundary to hear it.
  if (!isMobile) return;
  try {
    await invoke("set_system_bar_style", { light });
  } catch {
    // Decoration: a build without the plugin keeps white icons, which is what
    // it had before, and there is nothing a caller could do about it.
  }
}

File: crates/mumble-tauri/src/platform/android/system_bars.rs
//! Bridge to the Android `SystemBarsPlugin`, which keeps the page clear of
//! the status bar, the gesture bar and the keyboard, and sets the bars' icons
//! to read against whatever the page paints behind them.

use tauri::Wry;
use tauri::plugin::PluginHandle;

/// Managed state holding the Tauri mobile plugin handle for
/// `SystemBarsPlugin` (Kotlin).
pub struct SystemBarsHandle(pub PluginHandle<Wry>);

/// Dark bar icons when `light` (a light page behind them), light icons
/// otherwise. The plugin re-sends the insets with every call, so this is also
/// how a freshly loaded page gets them.
pub fn set_style(handle: &SystemBarsHandle, light: bool) -> Result<(), String> {
    handle
        .0
        .run_mobile_plugin::<()>("setStyle", serde_json::json!({ "light": light }))
        .map_err(|e| e.to_string())
}
